server/plugins/authentication.js

```diff
@@ -0,0 +1,148 @@
+const { makeExtendSchemaPlugin, gql } = require("graphile-utils");
+
+const PassportLoginPlugin = makeExtendSchemaPlugin(build => ({
+  typeDefs: gql`
+    input RegisterInput {
+      username: String!
+      email: String!
+      password: String!
+      name: String
+      avatarUrl: String
+    }
+
+    type RegisterPayload {
+      user: User! @pgField
+    }
+
+    input LoginInput {
+      username: String!
+      password: String!
+    }
+
+    type LoginPayload {
+      user: User! @pgField
+    }
+
+    extend type Mutation {
+      register(input: RegisterInput!): RegisterPayload
+      login(input: LoginInput!): LoginPayload
+    }
+  `,
+  resolvers: {
+    Mutation: {
+      async register(
+        mutation,
+        args,
+        context,
+        resolveInfo,
+        { selectGraphQLResultFromTable }
+      ) {
+        const { username, password, email, name, avatarUrl } = args.input;
+        const { rootPgPool, login, pgClient } = context;
+        try {
+          // Call our login function to find out if the username/password combination exists
+          const {
+            rows: [user],
+          } = await rootPgPool.query(
+            `select users.* from app_private.really_create_user(
+              username => $1,
+              email => $2,
+              email_is_verified => false,
+              name => $3,
+              avatar_url => $4,
+              password => $5
+            ) users where not (users is null)`,
+            [username, email, name, avatarUrl, password]
+          );
+
+          if (!user) {
+            throw new Error("Registration failed");
+          }
+
+          // Tell Passport.js we're logged in
+          await login(user);
```

Comment:
Maybe you can `await` these two promises together? That way, you can start the second one before the first completes.

```js
await Promise.all([
  // Tell Passport.js we're logged in
  login(user),
  // Tell pg we're logged in
  pgClient.query("select set_config($1, $2, true);", [
    "jwt.claims.user_id",
    user.id,
  ]),
]);
```

[Docs for `Promise.all()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/all)